### DISCUSSION TITLE
Types of operations which are only safe to be performed on the main thread?

### DISCUSSION THREAD

**Discussion by cxiao** [Vector35/binaryninja-api#3759] [API] [answered]
In the Python API, the [`mainthread` module](https://api.binary.ninja/binaryninja.mainthread-module.html) and several [classes in the `plugin` module](https://api.binary.ninja/binaryninja.plugin-module.html) offer functionality for performing work on background threads, as well as functionality for ensuring that certain operations are only run on the main thread (e.g. `mainthread.execute_on_main_thread`.

However, it is not clear what types of operations can only ever be safely executed on the main thread. The one I can think of immediately is operations that update the UI, but are there any others? For example:

- Is it safe to update settings (regardless of whether they are user, project, or resource scope) from a background thread?
- Is it safe to perform file operations on the underlying BNDB via the functionality exposed by [`filemetadata.FileMetadata`](https://api.binary.ninja/binaryninja.filemetadata-module.html?highlight=metadata#binaryninja.filemetadata.FileMetadata) from a background thread?

In IDA, the [IDA C++ SDK has the `THREAD_SAFE` keyword](https://www.hex-rays.com/products/ida/support/sdkdoc/pro_8h.html), which has the following description:

> `#define THREAD_SAFE`
> Functions callable from any thread are marked with this keyword.

[As of IDA 7.2](https://hex-rays.com/products/ida/news/7_2/), IDAPython enforces the following behaviour for functions _not_ marked with this keyword:

> IDAPython: all functions not marked as THREAD_SAFE in the C++ SDK, will now check that they are being called from the main thread, avoiding possible corruption or crashes

Does the Binary Ninja core already enforce that certain functions can only ever be called from the main thread, and if so is it possible to expose information about that in the API docs?

**Comment by plafosse** [comment, marked as answer]
We currently don't expose any of this information which is obviously problematic. I'm going to convert this discussion to a documentation issue.

**Reply by cxiao** [reply]
Thanks @plafosse! I saw you created #3769 for this, so I'll go and subscribe to that issue.
